src/java/org/apache/cassandra/io/sstable/format/bti/BtiTableReader.java

```diff
@@ -0,0 +1,558 @@
+        if (indexEntry == null)
+            return UnfilteredRowIterators.noRowsIterator(metadata(), key, Rows.EMPTY_STATIC_ROW, DeletionTime.LIVE, reversed);
+
+        if (reversed)
+            return new SSTableReversedIterator(this, dataFileInput, key, indexEntry, slices, selectedColumns, rowIndexFile);
+        else
+            return new SSTableIterator(this, dataFileInput, key, indexEntry, slices, selectedColumns, rowIndexFile);
+    }
+
+    @Override
+    public ISSTableScanner getScanner()
+    {
+        return BtiTableScanner.getScanner(this);
+    }
+
+    @Override
+    public ISSTableScanner getScanner(Collection<Range<Token>> ranges)
+    {
+        if (ranges != null)
+            return BtiTableScanner.getScanner(this, ranges);
+        else
+            return getScanner();
+    }
+
+    @Override
+    public ISSTableScanner getScanner(Iterator<AbstractBounds<PartitionPosition>> rangeIterator)
+    {
+        return BtiTableScanner.getScanner(this, rangeIterator);
+    }
+
+    @VisibleForTesting
+    @Override
+    public BtiTableReader cloneAndReplace(IFilter filter)
+    {
+        return unbuildTo(new Builder(descriptor).setFilter(filter), true).build(owner().orElse(null), true, true);
+    }
+
+    @Override
+    public BtiTableReader cloneWithRestoredStart(DecoratedKey restoredStart)
+    {
+        return runWithLock(ignored -> cloneAndReplace(restoredStart, OpenReason.NORMAL));
+    }
+
+    @Override
+    public BtiTableReader cloneWithNewStart(DecoratedKey newStart)
+    {
+        return runWithLock(d -> {
+            assert openReason != OpenReason.EARLY;
+            // TODO: merge with caller's firstKeyBeyond() work,to save time
+            if (newStart.compareTo(first) > 0)
+            {
+                final long dataStart = getPosition(newStart, Operator.EQ);
+                runOnClose(() -> dfile.dropPageCache(dataStart));
+            }
+
+            return cloneAndReplace(newStart, OpenReason.MOVED_START);
+        });
+    }
+
+    /**
+     * Clone this reader with the provided start and open reason, and set the clone as replacement.
+     *
+     * @param newFirst the first key for the replacement (which can be different from the original due to the pre-emptive
+     *                 opening of compaction results).
+     * @param reason   the {@code OpenReason} for the replacement.
+     * @return the cloned reader. That reader is set as a replacement by the method.
+     */
+    private BtiTableReader cloneAndReplace(DecoratedKey newFirst, OpenReason reason)
+    {
+        return unbuildTo(new Builder(descriptor), true)
+               .setFirst(newFirst)
+               .setOpenReason(reason)
+               .build(owner().orElse(null), true, true);
+    }
+
+    @Override
+    public DecoratedKey firstKeyBeyond(PartitionPosition token)
+    {
+        try
+        {
+            TrieIndexEntry pos = getRowIndexEntry(token, Operator.GT, true, false, SSTableReadsListener.NOOP_LISTENER);
```

Comment:
Just to confirm my understanding, `getRowIndexEntry()` produces a data file position in `pos.position` whether the partition is indexed or not?